libkbfs/conflict_resolver.go

```diff
@@ -119,40 +120,51 @@ func (cr *ConflictResolver) stopProcessing() {
 	cr.inputChan = nil
 }
 
+// cancelExistingLocked must be called while holding cr.inputLock.
+func (cr *ConflictResolver) cancelExistingLocked(ci conflictInput) bool {
+	// The input is only interesting if one of the revisions is
+	// greater than what we've looked at to date.
+	if ci.unmerged <= cr.currInput.unmerged &&
+		ci.merged <= cr.currInput.merged {
+		return false
+	}
+	if cr.currCancel != nil {
+		cr.currCancel()
+	}
+	return true
+}
+
 // processInput processes conflict resolution jobs from the given
 // channel until it is closed. This function uses a parameter for the
 // channel instead of accessing cr.inputChan directly so that it
 // doesn't have to hold inputChanLock.
 func (cr *ConflictResolver) processInput(baseCtx context.Context,
 	inputChan <-chan conflictInput) {
-	var cancel context.CancelFunc
 	var prevCRDone chan struct{}
 	defer func() {
-		if cancel != nil {
-			cancel()
+		cr.inputLock.Lock()
+		defer cr.inputLock.Unlock()
+		if cr.currCancel != nil {
+			cr.currCancel()
 		}
 		CleanupCancellationDelayer(baseCtx)
 	}()
 	for ci := range inputChan {
 		ctx := ctxWithRandomIDReplayable(baseCtx, CtxCRIDKey, CtxCROpID, cr.log)
 
-		valid := func() bool {
+		valid, doWait := func() (bool, bool) {
```

Comment:
you could get rid of `doWait` and simplify things a bit by:

1) initializing `prevCRDone` like:

```
// Start off with a closed prevCRDone.
prevCRDone := make(chan struct{})
close(prevCRDone)
```

Then you can just go:
```
waitChan := prevCRDone
prevCRDone = ...
```